cmds/df/df.go

```diff
@@ -0,0 +1,153 @@
+//
+// df reports details of mounted filesystems
+//
+// Synopsis
+//  df [-k] [-m]
+//
+// Description
+//  read mount information from /proc/mounts and
+//  statfs syscall and display summary information for all
+//  mount points that have a non-zero block count.
+//  Users can choose to see the diplay in KB or MB.
+//
+// Options
+//  -k: display values in KB (default)
+//  -m: dispaly values in MB
+//
+package main
+
+import (
+	"bytes"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"math"
+	"syscall"
+)
+
+var (
+	inKB  = flag.Bool("k", false, "Express the values in kilobytes (default)")
+	inMB  = flag.Bool("m", false, "Express the values in megabytes")
+	units uint64
+)
+
+const mountinfoFile = "/proc/mounts"
+
+const (
+	// B is Bytes
+	B = 1
+	// KB is kilobytes
+	KB = 1024 * B
+	// MB is megabytes
+	MB = 1024 * KB
+)
+
+// Mount is a structure used to contain mount point data
+type Mount struct {
+	Device         string
+	MountPoint     string
+	FileSystemType string
+	Flags          string
+	Bsize          int64
+	Blocks         uint64
+	Total          uint64
+	Used           uint64
+	Avail          uint64
+	PCT            uint8
+}
+
+type mountinfomap map[string]Mount
+
+// mountinfo returns a map of mounts representing
+// the data in /proc/mounts
+func mountinfo() (mountinfomap, error) {
+	buf, err := ioutil.ReadFile(mountinfoFile)
+	if err != nil {
+		return nil, err
+	}
+	return mountinfoFromBytes(buf)
+}
+
+// returns a map generated from the bytestream returned
+// from /proc/mounts
+// for tidiness, we decide to ignore filesystems of size 0
+func mountinfoFromBytes(buf []byte) (mountinfomap, error) {
+	ret := make(mountinfomap, 0)
+	for _, line := range bytes.Split(buf, []byte{'\n'}) {
+		kv := bytes.SplitN(line, []byte{' '}, 6)
+		if len(kv) != 6 {
+			// can't interpret this
+			continue
+		}
+		key := string(kv[1])
+		var mnt Mount
+		mnt.Device = string(kv[0])
+		mnt.MountPoint = string(kv[1])
+		mnt.FileSystemType = string(kv[2])
+		mnt.Flags = string(kv[3])
+		DiskUsage(&mnt)
+		if mnt.Blocks == 0 {
```

Comment:
Why ignore fs of size 0?

Reply:
Good question. cgroup, procfs and sysfs filesystems types, which are typically not reported by the df command are all neatly filtered by this method.